zproject/backends.py

```diff
@@ -156,13 +158,24 @@ def process_do_auth(self, user_profile, *args, **kwargs):
         if inactive_user or inactive_realm:
             return None
 
-        request = self.strategy.request  # type: ignore # This comes from Python Social Auth.
+        strategy = self.strategy  # type: ignore # This comes from Python Social Auth.
+        request = strategy.request
         email_address = self.get_email_address(*args, **kwargs)
         full_name = self.get_full_name(*args, **kwargs)
 
-        return login_or_register_remote_user(request, email_address,
-                                             user_profile, full_name,
-                                             bool(invalid_subdomain))
+        subdomain = strategy.session_get('subdomain')
+
+        if not subdomain:
+            return login_or_register_remote_user(request, email_address,
+                                                 user_profile, full_name,
+                                                 bool(invalid_subdomain))
+        else:
+            try:
+                realm = Realm.objects.get(string_id=subdomain)
+            except Realm.DoesNotExist:
+                return redirect_to_subdomain_login_url()
+
+        return redirect_and_log_into_subdomain(realm, full_name, email_address)
```

Comment:
This `if/else` logic should be cleaned up -- after a return statement we don't need the extra `else` statement.